lib/Reader/abisignature.cpp

```diff
@@ -86,43 +105,224 @@ ABICallSignature::ABICallSignature(const ReaderCallSignature &TheSignature,
+  }
+
+  // Push the unmanaged call frame
+  Value *FrameVPtr = getFieldAddress(Builder, CallFrame,
+                                     CallFrameInfo.offsetOfFrameVptr, Int8Ty);
+  Value *ThreadBase = Builder.CreateLoad(Thread);
+  Value *ThreadFrameAddress = getFieldAddress(
+      Builder, ThreadBase, JitContext.EEInfo.offsetOfThreadFrame, Int8PtrTy);
+  Builder.CreateStore(FrameVPtr, ThreadFrameAddress);
+
+  // Compute the address of the return address field
+  Value *ReturnAddressAddress = getFieldAddress(
+      Builder, CallFrame, CallFrameInfo.offsetOfReturnAddress, Int8PtrTy);
+
+  // Compute the address of the GC mode field
+  Value *GCStateAddress = getFieldAddress(
+      Builder, ThreadBase, JitContext.EEInfo.offsetOfGCState, Int8Ty);
+
+  // Compute address of the thread trap field
+  Value *ThreadTrapAddress = nullptr;
+  Type *ThreadTrapAddressTy = Reader.getUnmanagedPointerType(Int32Ty);
+  void *IndirectAddrOfCaptureThreadGlobal = nullptr;
+  void *AddrOfCaptureThreadGlobal =
+      (void *)JitContext.JitInfo->getAddrOfCaptureThreadGlobal(
+          &IndirectAddrOfCaptureThreadGlobal);
+  if (AddrOfCaptureThreadGlobal != nullptr) {
+    Value *RawThreadTrapAddress = ConstantInt::get(
+        LLVMContext, APInt(Reader.TargetPointerSizeInBits,
+                           (uint64_t)AddrOfCaptureThreadGlobal));
+    ThreadTrapAddress =
+        Builder.CreateIntToPtr(RawThreadTrapAddress, ThreadTrapAddressTy);
+  } else {
+    Value *IndirectThreadTrapAddress = ConstantInt::get(
+        LLVMContext, APInt(Reader.TargetPointerSizeInBits,
+                           (uint64_t)IndirectAddrOfCaptureThreadGlobal));
+    Type *IndirectAddressTy =
+        Reader.getUnmanagedPointerType(ThreadTrapAddressTy);
+    Value *TypedIndirectAddress =
+        Builder.CreateIntToPtr(IndirectThreadTrapAddress, IndirectAddressTy);
+    ThreadTrapAddress = Builder.CreateLoad(TypedIndirectAddress);
+  }
+
+  // Compute address of GC pause helper
+  Value *PauseHelperAddress =
+      (Value *)Reader.getHelperCallAddress(CORINFO_HELP_STOP_FOR_GC);
+
+  // Construct the call.
+  //
+  // The signature of the intrinsic is:
+  // @llvm.experimental_gc_transition(
+  //   fn_ptr target,
+  //   i32 numCallArgs,
+  //   i32 unused,
+  //   ... call args ...,
+  //   i32 numTransitionArgs,
+  //   ... transition args...,
+  //   i32 numDeoptArgs,
+  //   ... deopt args...)
+  //
+  // In the case of CoreCLR, there are 4 transition args and 0 deopt args.
+  //
+  // The transition args are:
+  // 0) Address of the return address field
+  // 1) Address of the GC mode field
+  // 2) Address of the thread trap global
+  // 3) Address of CORINFO_HELP_STOP_FOR_GC
+  Module *M = Reader.Function->getParent();
+  Type *CallTypeArgs[] = {Target->getType()};
+  Function *CallIntrinsic = Intrinsic::getDeclaration(
+      M, Intrinsic::experimental_gc_transition, CallTypeArgs);
+
+  const uint32_t PrefixArgCount = 3;
+  const uint32_t PostfixArgCount = 6;
```

Comment:
I would also add a symbolic const TransitionArgCount and use it here (+2) and down on line 248.